File: android/src/main/kotlin/com/app/rtmp_streaming/StreamUrlRedaction.kt
package com.app.rtmp_streaming

/**
 * Renders a publish URL safe to write to logcat.
 *
 * The stream key is the last path segment of an RTMP/RTMPS publish URL, and with
 * providers like Cloudflare Stream that segment *is* the whole credential --
 * anyone reading it can publish to the input. logcat is readable by adb and by
 * crash/telemetry collectors, so the raw URL must never be logged.
 *
 * Everything up to the last '/' is kept: scheme, host, port and app path are the
 * parts that actually matter when debugging a connect failure. The key itself is
 * masked down to its last four characters, which is enough to tell which key is
 * in use without being enough to publish with.
 *
 * Any query string is dropped wholesale rather than parsed -- some providers pass
 * auth tokens there, and dropping is the safe default for a shape we have not
 * enumerated.
 *
 * KNOWN LIMITATION: this covers only this plugin's own logging. RootEncoder's
 * CommandsManager logs the AMF releaseStream / FCPublish / publish commands, and
 * those carry the RTMP stream name -- which for Cloudflare Stream is the key. It
 * calls android.util.Log.i unconditionally and RtmpStreamClient.setLogs(false)
 * does not reach it (that only gates RtmpSender), so the key still reaches
 * logcat on every connect. Closing that needs a patched RootEncoder; it is a
 * deliberate open item, not an oversight.
 */
internal fun redactStreamUrl(url: String?): String {
    if (url.isNullOrEmpty()) return "(none)"

    val withoutQuery = url.substringBefore('?')
    val hadQuery = withoutQuery.length != url.length
    val cut = withoutQuery.lastIndexOf('/')

    // No path separator at all: we cannot tell URL from key, so keep none of it.
    if (cut < 0) return "<redacted>"

    val head = withoutQuery.substring(0, cut)
    val key = withoutQuery.substring(cut + 1)
    val masked = if (key.length <= 4) "****" else "****${key.takeLast(4)}"

    return head + "/" + masked + if (hadQuery) " (+query redacted)" else ""
}

File: ios/rtmp_streaming/Sources/StreamUrlRedaction.swift
import Foundation

/// Renders a publish URL safe to write to the device log.
///
/// The stream key is the last path segment of an RTMP/RTMPS publish URL, and with
/// providers like Cloudflare Stream that segment *is* the whole credential --
/// anyone reading it can publish to the input. Device logs are readable by other
/// tooling and by crash collectors, so the raw URL must never be printed.
///
/// Scheme, host, port and app path are kept because those are the parts that
/// matter when debugging a connect failure. The key is masked to its last four
/// characters: enough to tell which key is in use, not enough to publish with.
/// Any query string is dropped wholesale, since some providers pass auth tokens
/// there.
///
/// Mirrors `redactStreamUrl` in the Android source.
func redactStreamUrl(_ url: String?) -> String {
  guard let url = url, !url.isEmpty else { return "(none)" }

  let withoutQuery = url.components(separatedBy: "?")[0]
  let hadQuery = withoutQuery.count != url.count

  // No path separator at all: we cannot tell URL from key, so keep none of it.
  guard let cut = withoutQuery.lastIndex(of: "/") else { return "<redacted>" }

  let head = String(withoutQuery[withoutQuery.startIndex..<cut])
  let key = String(withoutQuery[withoutQuery.index(after: cut)...])
  let masked = key.count <= 4 ? "****" : "****\(key.suffix(4))"

  return head + "/" + masked + (hadQuery ? " (+query redacted)" : "")
}
